lib/kube/proxy/streamproto/proto.go

```diff
@@ -0,0 +1,269 @@
+			return nil, trace.Errorf("expected websocket control message, got %v", ty)
+		}
+
+		var msg metaMessage
+		if err := utils.FastUnmarshal(data, &msg); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if msg.ClientHandshake == nil {
+			return nil, trace.Errorf("expected websocket client handshake, got %v", msg)
+		}
+
+		s.Mode = msg.ClientHandshake.Mode
+	}
+
+	go s.readTask()
+	return s, nil
+}
+
+func (s *SessionStream) readTask() {
+	for {
+		terminated := false
+		defer s.closedC.Do(func() { close(s.CloseC) })
+		ty, data, err := s.conn.ReadMessage()
+		if err != nil {
+			return
+		}
+
+		if ty == websocket.BinaryMessage {
+			s.in <- data
+		}
+
+		if ty == websocket.TextMessage {
+			var msg metaMessage
+			if err := utils.FastUnmarshal(data, &msg); err != nil {
+				return
+			}
+
+			if msg.Resize != nil {
+				s.resizeQueue <- msg.Resize
+			}
+
+			if msg.ForceTerminate && !terminated {
+				terminated = true
+				close(s.forceTerminate)
+			}
+		}
+
+		if ty == websocket.CloseMessage {
+			s.conn.Close()
+			s.closed = true
+			return
+		}
+	}
+}
+
+func (s *SessionStream) Read(p []byte) (int, error) {
+	if len(s.currentIn) == 0 {
+		select {
+		case s.currentIn = <-s.in:
+		case <-s.CloseC:
+			return 0, io.EOF
+		}
+	}
+
+	n := copy(p, s.currentIn)
+	s.currentIn = s.currentIn[n:]
+	return n, nil
+}
+
+func (s *SessionStream) Write(data []byte) (int, error) {
+	s.writeSync.Lock()
+	defer s.writeSync.Unlock()
+
+	return len(data), s.conn.WriteMessage(websocket.BinaryMessage, data)
+}
+
+// Resize sends a resize request to the other party.
+func (s *SessionStream) Resize(size *remotecommand.TerminalSize) error {
+	msg := metaMessage{Resize: size}
+	json, err := utils.FastMarshal(msg)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.writeSync.Lock()
+	defer s.writeSync.Unlock()
+
+	return trace.Wrap(s.conn.WriteMessage(websocket.TextMessage, json))
+}
+
+// ResizeQueue returns a channel that will receive resize requests.
+func (s *SessionStream) ResizeQueue() chan *remotecommand.TerminalSize {
+	return s.resizeQueue
+}
+
+// ForceTerminate returns the channel used for force termination requests.
+func (s *SessionStream) ForceTerminate() chan struct{} {
+	return s.forceTerminate
+}
+
+// DoForceTerminate sends a force termination request to the other end.
+func (s *SessionStream) DoForceTerminate() error {
+	msg := metaMessage{ForceTerminate: true}
+	json, err := utils.FastMarshal(msg)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.writeSync.Lock()
+	defer s.writeSync.Unlock()
+
+	return trace.Wrap(s.conn.WriteMessage(websocket.TextMessage, json))
+}
+
+// WaitOnClose waits until the other end closes the stream.
+func (s *SessionStream) WaitOnClose() {
+	<-s.CloseC
+}
+
+// Close closes the stream.
+func (s *SessionStream) Close() error {
+	if !s.closed {
+		s.conn.WriteMessage(websocket.CloseMessage, []byte{})
+
+		select {
+		case <-s.CloseC:
+		case <-time.After(time.Second * 5):
```

Comment:
- Should the timeout be considered an error?
- What sets `s.closed`?

Reply:
Ended up reworking this so it's set properly now. It's to prevent a double closure.